src/overmap.cpp

```diff
@@ -4244,95 +4227,63 @@ void overmap::place_specials_pass( std::vector<std::pair<const overmap_special *
-        const std::string unplaced = enumerate_as_string( mandatory.begin(), mandatory.end(),
-        []( const std::pair<const overmap_special *, int> &elem ) {
-            return string_format( "%s (%d)", elem.first->id.c_str(), elem.second );
-        } );
-        dbg( D_WARNING ) << string_format( "couldn't place mandatory overmap specials during first pass (with city checks): %s.", unplaced.c_str() );
-
-        // Couldn't place mandatory specials, try without city checks
-        place_specials_pass( mandatory, sectors, false );
-        current_validity = overmap_valid::unlimited;
-    }
-
-    place_specials_pass( optional, sectors, true );
-
-    unplaced_mandatory_specials.clear();
-    if( !mandatory.empty() ) {
-        current_validity = overmap_valid::invalid;
-    }
-
-    const auto &unplaced = mandatory.empty() ? mandatory_unlimited : mandatory;
-    if( !unplaced.empty() ) {
-        const std::string unplaced_s = enumerate_as_string( mandatory.begin(), mandatory.end(),
-        []( const std::pair<const overmap_special *, int> &elem ) {
-            return string_format( "%s (%d)", elem.first->id.c_str(), elem.second );
-        } );
-        dbg( D_WARNING ) << string_format( "couldn't place mandatory overmap specials: %s.", unplaced_s.c_str() );
-        std::transform( unplaced.begin(), unplaced.end(), std::back_inserter( unplaced_mandatory_specials ),
-        []( const std::pair<const overmap_special *, int> &pr ) {
-            return pr.first;
-        } );
+    // First insure that all minimum instance counts are met.
+    place_specials_pass( enabled_specials, sectors, false );
+    if( std::any_of( enabled_specials.begin(), enabled_specials.end(),
+                     []( overmap_special_placement placement ) {
+                         return placement.instances_placed < placement.special_details->occurrences.min;
+                     } ) ) {
+        // Randomly select from among the nearest uninitialized overmap positions.
+        int previous_distance = 0;
+        std::vector<point> nearest_candidates;
+        for( point candidate_addr : closest_points_first( 10, pos() ) ) {
+            if( !overmap_buffer.has( candidate_addr.x, candidate_addr.y ) ) {
+                int current_distance = square_dist( pos().x, pos().y, candidate_addr.x, candidate_addr.y );
+                if( nearest_candidates.empty() || current_distance == previous_distance ) {
+                    nearest_candidates.push_back( candidate_addr );
+                    previous_distance = current_distance;
+                } else {
+                    break;
+                }
+            }
+        }
+        if( !nearest_candidates.empty() ) {
+            std::random_shuffle( nearest_candidates.begin(), nearest_candidates.end() );
+            point new_om_addr = nearest_candidates.front();
+            overmap_buffer.create_custom_overmap( new_om_addr.x, new_om_addr.y, enabled_specials );
+        }
```

Comment:
"Must generate or print an error"
If an object fails to generate within 2 overmap radius from the overmap where it was supposed to be, it must print a warning.

It is a big regression if there is no such warning.
It's not acceptable to generate all overmaps until it fails. Is there a proper end condition or will it only fail when stack depth is exceeded or writing a file fails? I'm not sure if I'm reading this right, but it looks to me like it will recursively push the excess specials around for as long as it can.
If it keeps generating overmaps indefinitely, there is no way of checking if the settings/specials are good enough to actually ever produce an overmap.

It would wreck the ability to test extensive mods (say, More Locations, that ships with core) and also make it much harder to generate a world with "too many" (hard to tell what is "too many" without manually testing) mods that add new specials.
Even worse if we add custom settings, which don't even warn that the world gen might fail.

I see the possible solutions:
* Pass around stack depth (as in, number of `generate_custom_overmap` above it) to the generating function. Fail if it exceeds some value, preferably configurable, so that players with weaker computers don't wait all day to get that failure message.
* With each special, pass the overmap coord of the overmap where this special was supposed to spawn (first such overmap). If current failed overmapgen is >2 overmap radius from any special's parent overmap, fail generation.

Also, it would be more "robust" if the `generate_custom_overmap` tried to cluster the new overmaps around the starting point instead of always going in one direction (up, I think).